### PR TITLE
Org country modifiable

### PR DESCRIPTION
**Pull request by maggarwal13**

### CODE REVIEW

File: whois-endtoend/src/test/groovy/net/ripe/db/whois/spec/update/OrgSpec.groovy [hunk 1]
```diff
@@ -3039,16 +3039,12 @@ class OrgSpec extends BaseQueryUpdateSpec {
 
         then:
         def ack = ackFor message
-        ack.failed
```
Comment (eshryane) [inline whois-endtoend/src/test/groovy/net/ripe/db/whois/spec/update/OrgSpec.groovy:3042]: Please add new tests to add, modify, remove "country:" code on an organisation *with* resources maintained by the RIPE NCC, this should still not be allowed.

File: whois-endtoend/src/test/groovy/net/ripe/db/whois/spec/update/OrgSpec.groovy [hunk 2]
```diff
@@ -3652,4 +3640,48 @@ class OrgSpec extends BaseQueryUpdateSpec {
                 [ "Country not recognised: FF" ]
     }
 
+    def "modify organisation, org-type:OTHER, change country code with user password"() {
```
Comment (eshryane) [inline whois-endtoend/src/test/groovy/net/ripe/db/whois/spec/update/OrgSpec.groovy:3643]: Is there also a test for create organisation with country code? This should always be allowed.

File: whois-update/src/main/java/net/ripe/db/whois/update/domain/UpdateMessages.java [hunk 3]
```diff
@@ -195,6 +195,11 @@ public static Message cantChangeOrgName() {
                 "Please contact \"ncc@ripe.net\" to change the name.");
     }
 
+    public static Message cantChangeCountryCode() {
```
Comment (eshryane) [inline whois-update/src/main/java/net/ripe/db/whois/update/domain/UpdateMessages.java:198]: Why not re-use the existing error message:
`Attribute \"country:\" can only be changed by the RIPE NCC for this object. Please contact \"ncc@ripe.net\" to change it."`

File: whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/organisation/OrgNameAndCountryAttrValidator.java [hunk 4]
```diff
@@ -26,7 +27,7 @@
 import java.util.Set;
 
 @Component
-public class OrgNameNotChangedValidator implements BusinessRuleValidator {
+public class OrgNameAndCountryAttrValidator implements BusinessRuleValidator {
```
[the remark this answers is not in the document]
Comment (eshryane) [reply whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/organisation/OrgNameAndCountryAttrValidator.java:30]: OK

File: whois-endtoend/src/test/groovy/net/ripe/db/whois/spec/update/OrgSpec.groovy (hunk 1 above)
Comment (maggarwal13) [reply whois-endtoend/src/test/groovy/net/ripe/db/whois/spec/update/OrgSpec.groovy:3042]: Added 3 more tests

File: whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/organisation/OrgNameAndCountryAttrValidator.java [hunk 5]
```diff
@@ -84,14 +92,19 @@ private boolean isReferencedByRsMaintainedResource(final RpslObject rpslObject)
...
+        final CIString updatedAttrValue = updatedObject.getValueOrNullForAttribute(attrType);
 
-    private boolean orgNameDidntChange(final RpslObject originalObject, final RpslObject updatedObject) {
-        final CIString originalOrgName = originalObject.getValueOrNullForAttribute(AttributeType.ORG_NAME);
-        final CIString updatedOrgName = updatedObject.getValueOrNullForAttribute(AttributeType.ORG_NAME);
+        if(originalAttrValue == null && updatedAttrValue == null) {
+            return false;
+        }
+
+        if(originalAttrValue == null || updatedAttrValue == null) {
```
Comment (eshryane) [inline whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/organisation/OrgNameAndCountryAttrValidator.java:103]: Remove duplicate code

File: whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/organisation/OrgNameAndCountryAttrValidator.java [hunk 6]
```diff
@@ -84,14 +92,19 @@ private boolean isReferencedByRsMaintainedResource(final RpslObject rpslObject)
...
+            return true;
+        }
 
-        return (originalOrgName != null) &&
-                (updatedOrgName != null) &&
-                (Objects.equals(originalOrgName.toString(), updatedOrgName.toString()));
+        return !(Objects.equals(originalAttrValue.toString(), updatedAttrValue.toString()));
```
Comment (eshryane) [inline whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/organisation/OrgNameAndCountryAttrValidator.java:107]: Or, restore the original logic, I think it's more readable
`
return (originalAttrValue != null) &&
(updatedAttrValue != null) &&
(Objects.equals(originalAttrValue.toString(), updatedAttrValue.toString()));
`

File: whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/organisation/OrgNameAndCountryAttrValidator.java (hunk 5 above)
Comment (maggarwal13) [reply whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/organisation/OrgNameAndCountryAttrValidator.java:103]: May be a better way to handle this?

File: whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/organisation/OrgNameAndCountryAttrValidator.java (hunk 6 above)
Comment (maggarwal13) [reply whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/organisation/OrgNameAndCountryAttrValidator.java:107]: Old logic will not work as org-name can never be null, it was not handling that case

Comment (eshryane) [reply whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/organisation/OrgNameAndCountryAttrValidator.java:107]: OK

File: whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/organisation/OrgNameAndCountryAttrValidator.java (hunk 5 above)
Comment (maggarwal13) [reply whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/organisation/OrgNameAndCountryAttrValidator.java]: Simplified the logic, CISTRING tostring method was the cause of null pointer exception

Comment (eshryane) [reply whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/organisation/OrgNameAndCountryAttrValidator.java]: You can simplify the logic like:
```
        return ((originalAttrValue != null && updatedAttrValue == null) ||     // removed
                    (originalAttrValue == null && updatedAttrValue != null) ||.    // added
                    !Objects.equals(originalAttrValue, updatedAttrValue));         // changed
```
which I confirmed with these cases before and afterwards (although the method is private):
```
    @Test
    public void test() {
        assertThat(isAttributeModified(AttributeType.COUNTRY, RpslObject.parse("organisation: ORG-RIEN1-RIPE"), RpslObject.parse("organisation: ORG-RIEN1-RIPE")), is(false));       // not present
        assertThat(isAttributeModified(AttributeType.COUNTRY, RpslObject.parse("organisation: ORG-RIEN1-RIPE"), RpslObject.parse("organisation: ORG-RIEN1-RIPE\ncountry: NL")), is(true));   // added
        assertThat(isAttributeModified(AttributeType.COUNTRY, RpslObject.parse("organisation: ORG-RIEN1-RIPE\ncountry: NL"), RpslObject.parse("organisation: ORG-RIEN1-RIPE")), is(true));   // removed
        assertThat(isAttributeModified(AttributeType.COUNTRY, RpslObject.parse("organisation: ORG-RIEN1-RIPE\ncountry: NL"), RpslObject.parse("organisation: ORG-RIEN1-RIPE\ncountry: DE")), is(true));   // changed
        assertThat(isAttributeModified(AttributeType.COUNTRY, RpslObject.parse("organisation: ORG-RIEN1-RIPE\ncountry: NL"), RpslObject.parse("organisation: ORG-RIEN1-RIPE\ncountry: nl")), is(false));  // not changed
    }
```

Comment (eshryane) [reply whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/organisation/OrgNameAndCountryAttrValidator.java]: So it's a bit better, but not hugely, so feel free to leave it as-is (I was just confused by the separate if statements)

File: whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/organisation/OrgNameAndCountryAttrValidator.java [hunk 7]
```diff
@@ -84,14 +93,11 @@ private boolean isReferencedByRsMaintainedResource(final RpslObject rpslObject)
...
+        final String updatedAttrValue = updatedObject.containsAttribute(attrType) ? updatedObject.getValueForAttribute(attrType).toString() : null;
 
-    private boolean orgNameDidntChange(final RpslObject originalObject, final RpslObject updatedObject) {
-        final CIString originalOrgName = originalObject.getValueOrNullForAttribute(AttributeType.ORG_NAME);
-        final CIString updatedOrgName = updatedObject.getValueOrNullForAttribute(AttributeType.ORG_NAME);
-
-        return (originalOrgName != null) &&
-                (updatedOrgName != null) &&
-                (Objects.equals(originalOrgName.toString(), updatedOrgName.toString()));
+        return !StringUtils.equals(originalAttrValue, updatedAttrValue);
```
Comment (eshryane) [inline whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/organisation/OrgNameAndCountryAttrValidator.java:100]: Using Strings fails where the *case* of the value changes (e.g. NL to nl), but this *shouldn't* be considered a change, the value is case-insensitive.

Comment (eshryane) [reply whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/organisation/OrgNameAndCountryAttrValidator.java:100]: I think the old code had this issue also

**Pull request merged**